cpp/ql/src/semmle/code/cpp/ir/implementation/raw/IRSanity.qll

```diff
@@ -272,4 +273,18 @@ module InstructionSanity {
     func = switchInstr.getEnclosingIRFunction() and
     funcText = Language::getIdentityString(func.getFunction())
   }
+
+  query predicate invalidOverlap(
+    MemoryOperand useOperand, string message, IRFunction func, string funcText
+  ) {
+    exists(Overlap overlap |
+      overlap = useOperand.getDefinitionOverlap() and
+      overlap instanceof MayPartiallyOverlap and
+      message =
+        "MemoryOperand '" + useOperand.toString() + "' has a `getDefinitionOverlap()` of '" +
+          overlap.toString() + "'." and
+      func = useOperand.getEnclosingIRFunction() and
+      funcText = Language::getIdentityString(func.getFunction())
+    )
+  }
```

Comment:
In the raw IR (or in unaliased SSA, where aliased memory accesses are still unmodeled), every unmodeled use is hooked up to the result of `UnmodeledDefinition`, and is marked as `MustTotallyOverlapped`.

I'm not really happy with the existence of `UnmodeledDefinition` and `UnmodeledUse`. They exist because, early on, there was no `Operand` type. To access the operands of an `Instruction`, you used `Instruction Instruction.getOperand(OperandTag tag)`, which returned the definition of that operand. Without something like `UnmodeledDefinition`, you couldn't tell the difference between an operand that existed but was not modeled, vs. an operand that simply didn't exist at all.

`UnmodeledUse` exists so that unmodeled definitions don't appear to be "dead" when we really just don't know where they are used.

I'd like to get rid of both `UnmodeledDefinition` and `UnmodeledUse`. Neither has any use in aliased SSA IR, which is the only IR we expose publicly. In the earlier IR stages, I'd rather just have an `Instruction.isResultUnmodeled()` predicate and an `Operand.isUnmodeled()` predicate.

Reply:
> In the raw IR (or in unaliased SSA, where aliased memory accesses are still unmodeled), every unmodeled use is hooked up to the result of `UnmodeledDefinition`, and is marked as `MustTotallyOverlapped`.

Ah, I see. I was missing that part of the puzzle. Thank you!